### PR TITLE
Fix repo search result

### LINKED ISSUE #2595: After upgrading, repositories without an issuetracker can not be accessed.
- Gitea version (or commit ref): 1.2.0+rc3
- Git version: 2.11.0
- Operating system: Linux Debian 9.1 kernel 4.9.0-3-amd64
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:
routers/repo/view.go:274 Home()] [E] Home: Cannot find any unit on this repository which you are allowed to access
## Description
After upgrading from 1.1.4 to the release candidate (1.2.0+rc3), repositories without an issue tracker can not be accessed. You will get a 404 error, Page can not be found.

Switching back to 1.1.4 and enabling the internal issue tracker for that repository and switching again to 1.2.0+rc3 version, the repository is accessible again.

After disabling the issue tracker in 1.2.0+rc3 it is still accessible.
It stays accessible after switching back to 1.1.4 and 1.2.0+rc3 again.

This has been the case for rc1 and rc2 too.

### PR DESCRIPTION
**Pull request by daviian**
Targets #2562, #2399, #1794 and #2595 too.

fixes wrong search result.
Most of the changes are adding tests from @Morlinest

### CODE REVIEW

**Comment by daviian** [comment]
Should fix #2562 , #2399 , #1794 as well

**Comment by Morlinest** [comment]
First, you are using my (older) tests, while PR with my tests is still pending (#2575). Second, you had problem with review of my changes to repo search because PR was too big but now you are doing the same. I am dividing that PR to multiple PRs one by one, now blocked only by aprooving and merging... So I don't understand (and don't like) this.

**Comment by daviian** [comment]
@Morlinest As I told you in a PM in discord. Your tests are great. And instead of pushing a whole refactoring and fix I only did a fix. Your refactoring should be added after this and all is fine.
And I even said you can take my changes I did on thursday. (which are exactly this)

**Comment by Morlinest** [comment]
But why you are adding tests when I've already posted PR with tests only?

**Comment by daviian** [comment]
Because these are tests for the old code with `Searcher` as it is. And they are working.
The only difference is that they have the old search results. Like transitive ownership as described in your PR.
The code only removes non accessible repositories in list view,

**Comment by Morlinest** [comment]
@daviian That tests contain it too. I think you missed it:
```
if testCase.opts.OwnerID > 0 {
	testCase.opts.Searcher = &User{ID: testCase.opts.OwnerID}
}
```
It should be same (old vs new version), only rewritten to `testCases`.

**Comment by daviian** [comment]
@Morlinest It was important for me to leave existing code as it is. Just to make it more easier to see that nothing has changed. A rewriting of tests doesn't necessarily mean that tests haven't changed.

And as you can see it was important for me to give credits to you too.
And moreover this PR doesn't make your PR's useless. They should be added nevertheless as yours make code a lot cleaner.

**Comment by Morlinest** [comment]
I've created that tests, they are not in codebase, they have same results and improved. I haven't spent a little amount of time to do it. Nor first version of tests, nor second one.

**Comment by daviian** [comment]
@Morlinest Changed PR such that it totally uses your commit. And if that doesn't please you I can offer you to take my code from that PR and create your own PR. Or i remove your tests from this PR. What would you like?
I don't want anyone to be upset. Only to get 1.2 released.

File: integrations/api_repo_test.go [hunk 1]
```diff
@@ -46,6 +46,220 @@ func TestAPISearchRepoNotLogin(t *testing.T) {
 		assert.Contains(t, repo.Name, keyword)
 		assert.False(t, repo.Private)
 	}
+
+	// Should return all (max 50) public repositories
+	req = NewRequest(t, "GET", "/api/v1/repos/search?limit=50")
```
Comment (ethantkoenig) [inline integrations/api_repo_test.go:51]: This is a lot of duplicated code. Maybe we should add a helper function
```
func apiRepoSearchResult(t testing.T, params string, expectedLen int) *api.SearchResults
...
apiRepoSearchResult(t, "?limit=50", 12)
```

Even better, add a session parameter, so you can use it for the logged-in tests too.

File: models/org.go [hunk 2]
```diff
@@ -350,8 +350,11 @@ func GetOrgsByUserID(userID int64, showAll bool) ([]*User, error) {
 	return getOrgsByUserID(sess, userID, showAll)
 }
 
-func getOwnedOrgsByUserID(sess *xorm.Session, userID int64) ([]*User, error) {
+func getOwnedOrgsByUserID(sess *xorm.Session, userID int64, showAll bool) ([]*User, error) {
```
Comment (ethantkoenig) [inline models/org.go:353]: nit: could we call it `includePrivate` or something similar so it is clearer what is included and unincluded when the parameter is `false`?

File: models/repo_list_test.go [hunk 3]
```diff
@@ -60,10 +55,133 @@ func TestSearchRepositoryByName(t *testing.T) {
...
-		Searcher: &User{ID: 14},
 	})
 
-	assert.NotNil(t, repos)
 	assert.NoError(t, err)
 	assert.Equal(t, int64(3), count)
+	assert.Len(t, repos, 3)
+
+	// Get all public repositories by name
+	repos, count, err = SearchRepositoryByName(&SearchRepoOptions{
```
Comment (ethantkoenig) [inline models/repo_list_test.go:65]: This is a lot of duplicated code, maybe we should make this a [table-driven test](https://github.com/golang/go/wiki/TableDrivenTests).

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: integrations/api_repo_test.go (hunk 1 above)
Comment (daviian) [reply integrations/api_repo_test.go:51]: @ethantkoenig If you want that you can review https://github.com/go-gitea/gitea/pull/2575 as it rewrites tests. I wanted to change as few as possible to make reviewing easier, but if it is desired the mentioned PR does this ;-)

File: models/repo_list_test.go (hunk 3 above)
Comment (daviian) [reply models/repo_list_test.go:65]: Same as above ;-)

File: models/org.go (hunk 2 above)
Comment (daviian) [reply models/org.go:353]: I called it showAll to be consistent with other methods used similarly. But I can change that at least for my PR.

**Comment by Morlinest** [comment]
@daviian You know I have PR for solving problems with this function :). Anyway, regardless of tests. I looked at your changes to implementation. It looks like something I was doing in first version of my PR, see https://github.com/go-gitea/gitea/pull/2371/commits/dae998e50b9781db60d640eda40ae475c0333268. You are making same authorization decisions inside that shouldn't be there. I think more changes to current implementation with `Searcher` and `OwnerID` will not (can't) make it better.

My opinion:
Try to think about results. With your changes, if `Searcher` is not empty and you search for repositories of another user (`OwnerID`), you will get your own repositories and repositories of owned (or related) organizations + repositories owned by `OwnerID`. But if `Searcher` is empty, you will get repositories of `OwnerID` and repositories of organization owned by `OwnerID`. But I would expect to get just subset of repos from first scenario (when there is `Searcher` and `OwnerID`). In second scenario you will get some repositories from first case, but also some repositories that aren't returned in first scenario. Also it doesn't solve problem, when `Searcher` is set but `OwnerID` is not.

**Comment by daviian** [comment]
@Morlinest 
First of all, I don't want to change existing search behaviour, that is way out of scope. PR should only remove repositories a user has no access to. That's why I don't break down the whole search function.
And because it shouldn't be there you should nevertheless refactor that stuff in your PR after that.
Second as I commented in code: If `Searcher` is not empty it means that `OwnerID = SearcherID`

And last but not least and as I said multiple times: This PR doesn't replace your refactoring, which includes changing search behaviour. It is only an interim step to fix permission problems.

File: models/repo_list.go [hunk 4]
```diff
@@ -155,26 +155,44 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, coun
...
 				ownerIds = append(ownerIds, org.ID)
 			}
 
-			searcherReposCond = searcherReposCond.Or(builder.In("owner_id", ownerIds))
 			if opts.Collaborate {
-				searcherReposCond = searcherReposCond.Or(builder.Expr("id IN (SELECT repo_id FROM `access` WHERE access.user_id = ? AND owner_id != ?)",
+				searcherReposCond = searcherReposCond.Or(builder.Expr("id IN (SELECT repo_id FROM `access` WHERE access.user_id = ?) AND owner_id != ?",
 					opts.Searcher.ID, opts.Searcher.ID))
 			}
+		} else if opts.OwnerID > 0 {
```
Comment (Morlinest) [inline models/repo_list.go:176]: Not needed. It's already inside `opts.OwnerID > 0` block.

File: models/repo_list.go [hunk 5]
```diff
@@ -155,26 +155,44 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, coun
...
+			ownerIds = append(ownerIds, opts.OwnerID)
+
+			owner, err := GetUserByID(opts.OwnerID)
+			if err != nil {
+				return nil, 0, fmt.Errorf("User: %v", err)
+			}
+
+			err = owner.GetOwnedOrganizations(false)
```
Comment (Morlinest) [inline models/repo_list.go:184]: Authorization decision making.

File: models/repo_list.go [hunk 6]
```diff
@@ -155,26 +155,44 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, coun
...
+
+			for _, org := range owner.OwnedOrgs {
+				ownerIds = append(ownerIds, org.ID)
+			}
+
+			if err != nil {
+				return nil, 0, fmt.Errorf("Organization: %v", err)
+			}
+
```
Comment (Morlinest) [inline models/repo_list.go:193]: Missing `collaborate` repos.

File: models/repo_list.go [hunk 7]
```diff
@@ -155,26 +155,44 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, coun
...
 		}
+		searcherReposCond = searcherReposCond.Or(builder.In("owner_id", ownerIds))
```
Comment (Morlinest) [inline models/repo_list.go:195]: Already in condition, line 157.

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Comment by daviian** [comment]
Closing PR as there exist scenarios in which this only solves a small portion of the search problem.
So I see that we don't come around a not that small refactoring of search.

But I want to point out that search should be fixed asap it is a massive security issue IMO.

**Pull request closed without merging**